coverage/hare_coverage.py

```diff
@@ -0,0 +1,319 @@
+        except Exception as e:
+            logging.error(f"Error {e} while cleanup.")
+
+    def run_ut(self) -> None:
+        """
+        Execute unit testcases.
+        :return: None
+        """
+        logging.info("Running UT's.")
+        dir_scanner = DirectoryScanner("test")
+        for test_dir_path in dir_scanner.search_dir(getcwd()):
+            self.utils.run_cmd(f"pytest {test_dir_path}")
+
+    def run_coverage(self) -> None:
+        """
+        Running coverage tool over UTs.
+        :return: None
+        """
+        logging.info("Running Coverage.")
+        command_line: str
+        for index, (src, tst) in enumerate(self.src_tst_map.items()):
+            if index == 0:
+                command_line: str = f"pytest --cov={src} --cov-report={self.cov_type}:{self.report} {tst}"
+            else:
+                command_line: str = f"pytest --cov={src} --cov-report={self.cov_type}:{self.report} --cov-append {tst}"
+            self.utils.run_cmd(command_line)
+
+    def check_report(self) -> None:
+        """
+        Validate the report generated.
+        :return: None
+        """
+        logging.info("Checking report.")
+        if self.utils.is_exist(self.report) is True:
+            logging.info(f"Python Code Coverage Report generated successfully at {self.report!a}.")
+            print(f"Python Code Coverage Report generated successfully at {self.report!a}.")
+        else:
+            logging.error("Failed to generate the Coverage report.")
+            raise FailedToGenerateReport("Failed to generate the Coverage report.")
+
+
+class CCoverage(Strategy):
+
+    def cleanup(self) -> None:
+        """
+        Cleaning up old coverage reports.
+        :return: None
+        """
+        pass
+
+    def run_ut(self) -> None:
+        """
+        Execute unit testcases.
+        :return: None
+        """
+        pass
+
+    def run_coverage(self) -> None:
+        """
+        Running coverage tool over UTs.
+        :return: None
+        """
+        pass
+
+    def check_report(self) -> None:
+        """
+        Validate the report generated
+        :return: None
+        """
+        pass
+
+
+class Executor:
+
+    """
+    This class is responsible for the execution of the coverage for both
+    python and C.
+    """
+
+    def __init__(self, cov: Strategy, report_type: str):
+        """
+        Initialization.
+        :param cov: Coverage polymorphic class
+        :param report_type: type of the report xml/html
+        """
+        self.cov_instance = cov(report_type)
+
+    def execute(self) -> None:
+        """
+        Execute coverage helper functions.
+        :return: None
+        """
+        self.cov_instance.cleanup()
+        self.cov_instance.run_ut()
+        self.cov_instance.run_coverage()
+        self.cov_instance.check_report()
+
+
+if __name__ == '__main__':
+    Utility().is_inside_hare()
+    Executor(PythonCoverage, "xml").execute()

```

Comment:
mypy: Argument 1 to "Executor" has incompatible type "Type[PythonCoverage]"; expected "Strategy"

Reply:
Resolved